File: codes/cpp/chap07/answer_A49_greedy1_37454pts.cpp
// ###############################################
// # 책 XXX페이지 전반부의 탐욕 알고리즘을 이용한 구현입니다.
// ###############################################

#include <iostream>
using namespace std;

int T, P[109], Q[109], R[109];
int A[109];
int PatA[109], PatB[109];

int main() {
	// 입력
	cin >> T;
	for (int i = 1; i <= T; i++) cin >> P[i] >> Q[i] >> R[i];

	// 배열 A 초기화
	for (int i = 1; i <= 20; i++) A[i] = 0;

	// 탐욕 알고리즘
	int CurrentScore = 0;
	for (int i = 1; i <= T; i++) {
		// 패턴 A일 때의 점수 구하기
		int ScoreA = CurrentScore;
		for (int j = 1; j <= 20; j++) PatA[j] = A[j];
		PatA[P[i]] += 1;
		PatA[Q[i]] += 1;
		PatA[R[i]] += 1;
		for (int j = 1; j <= 20; j++) {
			if (PatA[j] == 0) ScoreA += 1;
		}

		// 패턴 B일 때의 점수 구하기
		int ScoreB = CurrentScore;
		for (int j = 1; j <= 20; j++) PatB[j] = A[j];
		PatB[P[i]] -= 1;
		PatB[Q[i]] -= 1;
		PatB[R[i]] -= 1;
		for (int j = 1; j <= 20; j++) {
			if (PatB[j] == 0) ScoreB += 1;
		}

		// 점수가 높은 쪽을 채택
		if (ScoreA >= ScoreB) {
			cout << "A" << endl;
			CurrentScore = ScoreA;
			for (int j = 1; j <= 20; j++) A[j] = PatA[j];
		}
		else {
			cout << "B" << endl;
			CurrentScore = ScoreB;
			for (int j = 1; j <= 20; j++) A[j] = PatB[j];
		}
	}
	return 0;
}

File: codes/python/chap07/answer_A49_greedy1_37454pts.py
################################################
# 책 XXX페이지 전반부의 탐욕 알고리즘을 이용한 구현입니다.
################################################

# 입력
T = int(input())
P = [ None ] * T
Q = [ None ] * T
R = [ None ] * T
for i in range(T):
	P[i], Q[i], R[i] = map(int, input().split())
	P[i] -= 1 # 0 다시 시작
	Q[i] -= 1
	R[i] -= 1

# 배열 A 초기화
A = [ 0 ] * 20

# 탐욕 알고리즘
CurrentScore = 0
for i in range(T):
	# 패턴 A일 때의 점수 구하기
	ScoreA = CurrentScore
	PatA = [ 0 ] * 20
	for j in range(20):
		PatA[j] = A[j]
	PatA[P[i]] += 1
	PatA[Q[i]] += 1
	PatA[R[i]] += 1
	for j in range(20):
		if PatA[j] == 0:
			ScoreA += 1

	# 패턴 B일 때의 점수 구하기
	ScoreB = CurrentScore
	PatB = [ 0 ] * 20
	for j in range(20):
		PatB[j] = A[j]
	PatB[P[i]] -= 1
	PatB[Q[i]] -= 1
	PatB[R[i]] -= 1
	for j in range(20):
		if PatB[j] == 0:
			ScoreB += 1

	# 점수가 높은 쪽을 채택
	if ScoreA >= ScoreB:
		print("A")
		CurrentScore = ScoreA
		for j in range(20):
			A[j] = PatA[j]
	else:
		print("B")
		CurrentScore = ScoreB
		for j in range(20):
			A[j] = PatB[j]
